makelib/imagelight.mk

```diff
@@ -0,0 +1,214 @@
+# Copyright 2021 The Upbound Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# ====================================================================================
+# Options
+
+ifeq ($(origin IMAGE_DIR),undefined)
+IMAGE_DIR := $(ROOT_DIR)/cluster/images
+endif
+
+ifeq ($(origin IMAGE_OUTPUT_DIR),undefined)
+IMAGE_OUTPUT_DIR := $(OUTPUT_DIR)/images/$(PLATFORM)
+endif
+
+ifeq ($(origin IMAGE_TEMP_DIR),undefined)
+IMAGE_TEMP_DIR := $(shell mktemp -d)
+endif
+
+# a registry that is scoped to the current build tree on this host. this enables
+# us to have isolation between concurrent builds on the same system, as in the case
+# of multiple working directories or on a CI system with multiple executors. All images
+# tagged with this build registry can safely be untagged/removed at the end of the build.
+ifeq ($(origin BUILD_REGISTRY), undefined)
+BUILD_REGISTRY := build-$(shell echo $(HOSTNAME)-$(ROOT_DIR) | shasum -a 256 | cut -c1-8)
+endif
+
+# In order to reduce built time especially on jenkins, we maintain a cache
+# of already built images. This cache contains images that can be used to help speed
+# future docker build commands using docker's content addressable schemes.
+# All cached images go in in a 'cache/' local registry and we follow an MRU caching
+# policy -- keeping images that have been referenced around and evicting images
+# that have to been referenced in a while (and according to a policy). Note we can
+# not rely on the image's .CreatedAt date since docker only updates then when the
+# image is created and not referenced. Instead we keep a date in the Tag.
+CACHE_REGISTRY := cache
+
+# prune images that are at least this many hours old
+PRUNE_HOURS ?= 48
```

Comment:
Couple of questions regarding the cache:
1. How do we plan to utilize the images that we're retagging to keep in the 'cache'?
I don't currently see anywhere where we're utilizing `cache-from` in the buildx commands? I see the `--load` arg being used [here](https://github.com/crossplane/crossplane/blob/e335e7c077d0269e6e1462eab44892c566e51be2/cluster/images/crossplane/Makefile#L17), however I think that's only effective for [single-platform builds](https://docs.docker.com/engine/reference/commandline/buildx_build/#load).

2. With `PRUNE_HOURS` set to 48, are there any concerns with the cache being only effective on subsequent branch builds, instead of being available for first build (for example when running on main/master)?
It seems like we'd have multiple items at play here:
- The build system (GH Actions, Circle, Travis, etc) and how it enables Docker build caching. As an example, Circle can be pretty hit or miss in what it has available in DLC. [ref](https://circleci.com/docs/2.0/docker-layer-caching/#how-dlc-works).
- How frequently builds are running. If a build only runs once every couple of days, it seems like we'll be starting from a "fresh" cache each time.

Reply:
@tnthornton these are all valid questions, to which my answer if mostly: We don't use the cache functionality here anywhere that I am aware of. The only reason it is included here is that it is 100% copied from the existing [image.mk](https://github.com/upbound/build/blob/master/makelib/image.mk), and I tried to keep everything that wouldn't actually prohibit us from using something like `buildx` (i.e. if the functionality is already present and it doesn't disrupt what we are trying to do then keep it available for consumers). That being said, I also considered stripping out everything that we don't need to make this as "light" as possible. Do you think you would advocate for that approach?

Reply:
@hasheddan sorry for the slow response.

>The only reason it is included here is that it is 100% copied from the existing image.mk, and I tried to keep everything that wouldn't actually prohibit us from using something like buildx (i.e. if the functionality is already present and it doesn't disrupt what we are trying to do then keep it available for consumers).

That makes sense to me and I'd probably take the same approach.

>That being said, I also considered stripping out everything that we don't need to make this as "light" as possible. Do you think you would advocate for that approach?

I think currently, yes(?). Especially if this was copied from the existing `image.mk`. Given that, if we find ourselves wanting to use cache pruning, etc we can bring it back or reference that impl.

FWIW, I do find myself going back and forth about this. So if one of the other reviewers find that they feel more strongly about just keeping it in here for now, I think I can be convinced to get behind that.

Reply:
I think I tend to agree -- part of the motivation for this change is our current process is "doing too much", so if we are not using something might as well remove it and not allow the potential for someone depending on it, then trying to remove it

Reply:
> Do you think you would advocate for that approach?

My intuition is also yes here. If this is going to be an alternative to the existing process it makes sense to me to take the opportunity to split out anything superfluous and 'start from scratch' rather than effectively forking what we already have an modifying it.

Reply:
Side note - if we do keep all the stuff that was copied from `image.mk` a diff between `image.mk` and `imagelight.mk` would help reviewers focus on what's new here (tough to do though since there's no `git cp` - maybe just a `git diff` in a comment? 😬)

Reply:
I have removed the caching / pruning / debug components, which I don't see us using anywhere of consequence. Also added a diff in the review at the bottom of the thread :)